File: src/test/com/natpryce/snodge/JsonBuilders.java
package com.natpryce.snodge;

import com.google.gson.*;
import com.natpryce.snodge.internal.JsonFunctions;

import java.util.Map;

import static com.natpryce.snodge.internal.JsonFunctions.entry;
import static java.util.Arrays.asList;

public class JsonBuilders {
    @SafeVarargs
	public static JsonObject object(Map.Entry<String, Object>... properties) {
		return toJsonObject(asList(properties));
	}

	public static Map.Entry<String, Object> withNullField(String name) {
		return entry(name, null);
	}

	public static Map.Entry<String, Object> withField(String name, String value) {
		return entry(name, (Object) value);
	}

	public static Map.Entry<String, Object> withField(String name, Integer value) {
		return entry(name, (Object) value);
	}

    public static Map.Entry<String, Object> withField(String name, JsonElement value) {
		return entry(name, (Object) value);
	}

    public static JsonArray list(Object... elements) {
		final JsonArray jsonArray = new JsonArray();
		for (Object element : elements)
		{
			jsonArray.add(asJsonElement(element));
		}
		return jsonArray;
	}

	private static JsonObject toJsonObject(Iterable<Map.Entry<String, Object>> entries)
	{
		JsonObject json = new JsonObject();
		for (Map.Entry<String, Object> property : entries) {
			json.add(property.getKey(), asJsonElement(property.getValue()));
		}
		return json;
	}

	private static JsonElement asJsonElement(Object element)
	{
		if (element == null) {
			return JsonNull.INSTANCE;
		}
		else if (element instanceof JsonElement) {
			return (JsonElement)element;
		}
		else if (element instanceof String) {
			return new JsonPrimitive((String)element);
		}
		else if (element instanceof Boolean) {
			return new JsonPrimitive((Boolean)element);
		}
		else if (element instanceof Number) {
			return new JsonPrimitive((Number)element);
		}
		else if (element instanceof Character) {
			return new JsonPrimitive((Character)element);
		}
		else {
			throw new IllegalArgumentException("cannot turn a " + element.getClass().getName() + " to a JsonElement");
		}
	}
}

File: src/test/com/natpryce/snodge/JsonBuilders.kt
@file:JvmName("JsonBuilders")

package com.natpryce.snodge

import com.google.gson.*
import com.natpryce.snodge.internal.JsonFunctions

import com.natpryce.snodge.internal.JsonFunctions.entry
import java.util.Arrays.asList

@SafeVarargs
fun obj(vararg properties: Map.Entry<String, Any>): JsonObject {
    return toJsonObject(asList<Map.Entry<String, Any>>(*properties))
}

fun withNullField(name: String): Map.Entry<String, Any> {
    return entry<String, Any>(name, null)
}

fun withField(name: String, value: String): Map.Entry<String, Any> {
    return entry(name, value as Any)
}

fun withField(name: String, value: Int?): Map.Entry<String, Any> {
    return entry<String, Any>(name, value as Any?)
}

fun withField(name: String, value: JsonElement): Map.Entry<String, Any> {
    return entry(name, value as Any)
}

fun list(vararg elements: Any?): JsonArray {
    val jsonArray = JsonArray()
    for (element in elements) {
        jsonArray.add(asJsonElement(element))
    }
    return jsonArray
}

private fun toJsonObject(entries: Iterable<Map.Entry<String, Any>>): JsonObject {
    val json = JsonObject()
    for (property in entries) {
        json.add(property.key, asJsonElement(property.value))
    }
    return json
}

private fun asJsonElement(element: Any?): JsonElement {
    if (element == null) {
        return JsonNull.INSTANCE
    }
    else if (element is JsonElement) {
        return element
    }
    else if (element is String) {
        return JsonPrimitive((element as String?)!!)
    }
    else if (element is Boolean) {
        return JsonPrimitive((element as Boolean?)!!)
    }
    else if (element is Number) {
        return JsonPrimitive((element as Number?)!!)
    }
    else if (element is Char) {
        return JsonPrimitive((element as Char?)!!)
    }
    else {
        throw IllegalArgumentException("cannot turn a " + element.javaClass.name + " to a JsonElement")
    }
}
